C error on appending aliased fixed arrays into array of sumtype

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

**V doctor:**

```
V full version: V 0.4.8 5ec9bb5.f2c281e
OS: linux, Linux version 6.6.8-arch1-1 (linux@archlinux) (gcc (GCC) 13.2.1 20230801, GNU ld (GNU Binutils) 2.41.0) #1 SMP PREEMPT_DYNAMIC Thu, 21 Dec 2023 19:01:01 +0000
Processor: 16 cpus, 64bit, little endian, 12th Gen Intel(R) Core(TM) i5-1240P

getwd: /home/ge/Code/v/netaddr
vexe: /home/ge/.vlang/v
vexe mtime: 2024-11-29 19:48:59

vroot: OK, value: /home/ge/.vlang
VMODULES: OK, value: /home/ge/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.46.1
Git vroot status: weekly.2024.47-62-gf2c281ea
.git/config present: true

CC version: cc (GCC) 14.2.1 20240910
emcc version: N/A
thirdparty/tcc status: thirdparty-linux-amd64 0134e9b9
```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg aliased_fixed_arrays_array.v`

```v
module main

type IPv4 = [4]u8
type IPv6 = [16]u8
type IP = IPv4 | IPv6

fn main() {
	mut arr := []IP{}
	arr << IPv4([4]u8{})
	arr << IPv6([16]u8{})
	println(arr)
}
```

**What did you expect to see?**

Something like `[IP([0, 0, 0, 0]), IP([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])]` or some V error since the following variant works (but not as expected, see output):

```v
module main

type IPv4 = [4]u8
type IPv6 = [16]u8
type IP = IPv4 | IPv6

fn main() {
	mut arr := []IP{}
	arr << IP(IPv4([4]u8{}))
	arr << IP(IPv6([16]u8{}))
	println(arr)
}
// Output: [IP(IPv4([72, 0, 0, 0])), IP(IPv6([248, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]))]
```

Note: only zeros is expected in output, but fixed arrays have random non-zero first item. This is also bug.

**What did you see instead?**

```
[unknown sum type value, unknown sum type value]
```

C error will shown only if `-show-c-output` is passed:

```
$ v -show-c-output run aliased_fixed_arrays_array.v
======== Output of the C Compiler (/home/ge/.vlang/thirdparty/tcc/tcc.exe) ========
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13871: error: too many initializers
===================================================================================
======== Output of the C Compiler (cc) ========
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c: In function ‘main__main’:
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13871:60: warning: excess elements in struct initializer
13871 |         array_push((array*)&arr, _MOV((main__IP[]){ {0, 0, 0, 0} }));
      |                                                            ^
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13871:60: note: (near initialization for ‘(anonymous)[0]’)
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13871:63: warning: excess elements in struct initializer
13871 |         array_push((array*)&arr, _MOV((main__IP[]){ {0, 0, 0, 0} }));
      |                                                               ^
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13871:63: note: (near initialization for ‘(anonymous)[0]’)
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:60: warning: excess elements in struct initializer
13872 |         array_push((array*)&arr, _MOV((main__IP[]){ {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0} }));
      |                                                            ^
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:60: note: (near initialization for ‘(anonymous)[0]’)
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:63: warning: excess elements in struct initializer
13872 |         array_push((array*)&arr, _MOV((main__IP[]){ {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0} }));
      |                                                               ^
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:63: note: (near initialization for ‘(anonymous)[0]’)
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:66: warning: excess elements in struct initializer
13872 |         array_push((array*)&arr, _MOV((main__IP[]){ {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0} }));
      |                                                                  ^
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:66: note: (near initialization for ‘(anonymous)[0]’)
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:69: warning: excess elements in struct initializer
13872 |         array_push((array*)&arr, _MOV((main__IP[]){ {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0} }));
      |                                                                     ^
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:69: note: (near initialization for ‘(anonymous)[0]’)
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:72: warning: excess elements in struct initializer
13872 |         array_push((array*)&arr, _MOV((main__IP[]){ {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0} }));
      |                                                                        ^
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:72: note: (near initialization for ‘(anonymous)[0]’)
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:75: warning: excess elements in struct initializer
13872 |         array_push((array*)&arr, _MOV((main__IP[]){ {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0} }));
      |                                                                           ^
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:75: note: (near initialization for ‘(anonymous)[0]’)
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:78: warning: excess elements in struct initializer
13872 |         array_push((array*)&arr, _MOV((main__IP[]){ {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0} }));
      |                                                                              ^
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:78: note: (near initialization for ‘(anonymous)[0]’)
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:81: warning: excess elements in struct initializer
13872 |         array_push((array*)&arr, _MOV((main__IP[]){ {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0} }));
      |                                                                                 ^
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:81: note: (near initialization for ‘(anonymous)[0]’)
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:84: warning: excess elements in struct initializer
13872 |         array_push((array*)&arr, _MOV((main__IP[]){ {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0} }));
      |                                                                                    ^
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:84: note: (near initialization for ‘(anonymous)[0]’)
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:87: warning: excess elements in struct initializer
13872 |         array_push((array*)&arr, _MOV((main__IP[]){ {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0} }));
      |                                                                                       ^
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:87: note: (near initialization for ‘(anonymous)[0]’)
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:90: warning: excess elements in struct initializer
13872 |         array_push((array*)&arr, _MOV((main__IP[]){ {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0} }));
      |                                                                                          ^
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:90: note: (near initialization for ‘(anonymous)[0]’)
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:93: warning: excess elements in struct initializer
13872 |         array_push((array*)&arr, _MOV((main__IP[]){ {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0} }));
      |                                                                                             ^
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:93: note: (near initialization for ‘(anonymous)[0]’)
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:96: warning: excess elements in struct initializer
13872 |         array_push((array*)&arr, _MOV((main__IP[]){ {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0} }));
      |                                                                                                ^
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:96: note: (near initialization for ‘(anonymous)[0]’)
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:99: warning: excess elements in struct initializer
13872 |         array_push((array*)&arr, _MOV((main__IP[]){ {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0} }));
      |                                                                                                   ^
/tmp/v_1000/aliased_fixed_arrays_array.01JDWR1SZD1QKEY4X09ETJSN4V.tmp.c:13872:99: note: (near initialization for ‘(anonymous)[0]’)
===============================================
[unknown sum type value, unknown sum type value]
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzRhMWZmYmQyZWY0Y2JjYzQ2Y2FiOTkiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.qoK8oB2i6SUPWHzhYW4Tf6idg3vK-R1cRa_oqwLNeus">Huly&reg;: <b>V_0.6-21450</b></a></sub>

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2695,8 +2695,9 @@ fn (mut g Gen) call_cfn_for_casting_expr(fname string, expr ast.Expr, exp_is_ptr
 	}
 	g.write('${fname}(')
 	if !got_is_ptr && !got_is_fn {
-		if is_comptime_variant || !expr.is_lvalue()
-			|| (expr is ast.Ident && expr.obj.is_simple_define_const()) {
+		is_cast_array_init := expr is ast.CastExpr && expr.expr is ast.ArrayInit
+		if !is_cast_array_init && (is_comptime_variant || !expr.is_lvalue()
+			|| (expr is ast.Ident && expr.obj.is_simple_define_const())) {
 			// Note: the `_to_sumtype_` family of functions do call memdup internally, making
 			// another duplicate with the HEAP macro is redundant, so use ADDR instead:
 			if expr.is_as_cast() {

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2695,8 +2695,9 @@ fn (mut g Gen) call_cfn_for_casting_expr(fname string, expr ast.Expr, exp_is_ptr
 	}
 	g.write('${fname}(')
 	if !got_is_ptr && !got_is_fn {
-		is_cast_array_init := expr is ast.CastExpr && expr.expr is ast.ArrayInit
-		if !is_cast_array_init && (is_comptime_variant || !expr.is_lvalue()
+		is_cast_fixed_array_init := expr is ast.CastExpr
+			&& (expr.expr is ast.ArrayInit && expr.expr.is_fixed)
+		if !is_cast_fixed_array_init && (is_comptime_variant || !expr.is_lvalue()
 			|| (expr is ast.Ident && expr.obj.is_simple_define_const())) {
 			// Note: the `_to_sumtype_` family of functions do call memdup internally, making
 			// another duplicate with the HEAP macro is redundant, so use ADDR instead:

--- a/vlib/v/tests/aliases/alias_array_fixed_sumtype_test.v
+++ b/vlib/v/tests/aliases/alias_array_fixed_sumtype_test.v
@@ -1,0 +1,12 @@
+module main
+
+type IPv4 = [4]u8
+type IPv6 = [16]u8
+type IP = IPv4 | IPv6
+
+fn test_main() {
+	mut arr := []IP{}
+	arr << IP(IPv4([4]u8{}))
+	arr << IP(IPv6([16]u8{}))
+	assert arr.str() == '[IP(IPv4([0, 0, 0, 0])), IP(IPv6([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]))]'
+}

